Add Arbitrary Batched Verifiable Tokens (#13)
ietf-wg-privacypass/ietf-draft-privacypass-batched-tokens: draft-ietf-privacypass-batched-tokens.md

which can either be publicly verifiable or not. While it is possible to run
multiple instances of the issuance protocol in parallel, e.g., over a
multiplexed transport such as HTTP/3 {{?HTTP3=RFC9114}} or by orchestrating multiple
HTTP requests, these ad-hoc solutions vary based on transport protocol support. In
addition, in some cases, they cannot take advantage of cryptographic optimizations.

The first variant of the issuance protocol builds upon the privately verifiable
issuance protocol in {{RFC9578}} that uses VOPRF {{!OPRF=I-D.irtf-cfrg-voprf}},
and allows for batched issuance of tokens. This allows clients to request more
than one token at a time and for issuers to issue more than one token at a
time. In effect, private batched issuance performance scales better than linearly.

The second variant of the issuance protocol introduces a new Client-Issuer communication method,
which allows for batched issuance of arbitrary token types. This allows clients to request more
than one token at a time and for issuers to issue more than one token at a
time. In effect, arbitrary batched tokens performance is linear.

> This batched issuance protocol registers two new token types ({{iana-token-type}}),

Author: ```suggestion
This batched issuance protocol registers one new token type ({{iana-token-type}}),
```

which can either be publicly verifiable or not. While it is possible to run
multiple instances of the issuance protocol in parallel, e.g., over a
multiplexed transport such as HTTP/3 {{?HTTP3=RFC9114}} or by orchestrating multiple
HTTP requests, these ad-hoc solutions varies from one implementation to the next. In
addition, they cannot take advantage of cryptographic optimisations, and their cost
always scales linearly with the number of instances.

The first variant of the issuance protocol builds upon the privately verifiable
issuance protocol in {{ISSUANCE}} that uses VOPRF {{!OPRF=I-D.irtf-cfrg-voprf}},
and allows for batched issuance of tokens. This allows clients to request more
than one token at a time and for issuers to issue more than one token at a
time. In effect, private batched issuance performance scales better than linearly.

The second variant of the issuance protocol introduces a new Client-Issuer communication method,
which allows for batched issuance of arbitrary token types. This allows clients to request more
than one token at a time and for issuers to issue more than one token at a
time. In effect, arbitrary batched tokens performance is linear.

> This batched issuance protocol registers two new token types ({{iana-token-type}}),

Author: you're right. thanks!

1. Improving server and client issuance efficiency by amortizing the cost of the
   VOPRF proof generation and verification, respectively.

For the Publicly Verifiable Token issuance protocol, it allows for a single TokenRequest

> to be sent that encompasses multiple tokens. The cost remains linear.

Reviewer: This is also true for the arbitrary batched issuance variant with private tokens

# Batched Privately Verifiable Token

This section describes a batched issuance protocol for select token types, including 0x0001 (defined in {{RFC9578}}) and 0xF91A (defined in this document).

> 

Reviewer: We need to note that this variant is primarily meant to be more efficient, it comes at the cost of losing the flexibility to choose a different truncated_token_key_id for every token.

1. Improving server and client issuance efficiency by amortizing the cost of the
   VOPRF proof generation and verification, respectively.

For the Publicly Verifiable Token issuance protocol, it allows for a single TokenRequest

> to be sent ........

Author: missed this one. updated.

~~~

If the FinalizeBatch function fails, the Client aborts the protocol. Token
verification works exactly as specified in {{RFC9578}}.

# Arbitrary Batched Token Issuance

This section describes an issuance protocol mechanism for issuing multiple tokens in one round trip between Client and Issuer. An arbitrary batched token request can contain token requests for any token type.

## Client-to-Issuer Request {#arbitrary-client-to-issuer-request}

The Client first creates all TokenRequest it wants to batch. To do so, the client follows
protocol describing issuance, such as {{RFC9578, Section 5.1}} or {{RFC9578, Section 6.1}}.

The Client then creates a BatchedTokenRequest structured as follows:

~~~tls
struct {

>   uint16_t token_type;
>   opaque token_request<0..2^16-1>;
> } TokenRequest

Reviewer 2: ```suggestion
struct {
   uint16_t token_type;
   select (token_type) {
      case (0x0001): /* Type VOPRF(P-384, SHA-384), RFC 9578 */
         uint8_t truncated_token_key_id;
         uint8_t blinded_msg[Ne];
      case (0x0002): /* Type Blind RSA (2048-bit), RFC 9578 */
         uint8_t truncated_token_key_id;
         uint8_t blinded_msg[Nk];
   }
} TokenRequest;
```
This isn't quite right. I think we just need to say that `TokenRequest` is simply prefixed by a uint16 value indicating the type, and that the rest of the structure follows based on that type.

Author: I've added this as an example. I have the impression the `opaque` structure better capture the general case, and the applicability to current and future token types

Reviewer 2: `opaque token_request<0..2^16-1>;` means that the field `token_request` is length-prefix-encoded with two bytes, which is not how TokenRequest messages are structured. So it wasn't a matter of clarity, it just didn't match what's in the actual spec.

The Client then generates an HTTP POST request to send to the Issuer Request
URL, with the BatchTokenRequest as the content. The media type for this request
is "application/private-token-batch-request". An example request for the Issuer
Request URL "https://issuer.example.net/request" is shown below.

~~~
POST /request HTTP/1.1
Host: issuer.example.net
Accept: application/private-token-batch-response
Content-Type: application/private-token-batch-request
Content-Length: <Length of BatchTokenRequest>

<Bytes containing the BatchTokenRequest>
~~~

## Issuer-to-Client Response {#arbitrary-issuer-to-client-response}

Upon receipt of the request, the Issuer validates the following conditions:

- The Content-Type is application/private-token-batch-request as registered with IANA.

If this condition is not met, the Issuer MUST return an HTTP 400 error
to the client.

> The Issuer then tries to deserialize the first 4 bytes of the i-th element

Reviewer 2: ```suggestion
The Issuer then tries to deserialize the first 2 bytes of the i-th element
```

Author: updated

of BatchTokenRequest.token_requests.
If this is not a token type registered with IANA, the Issuer MUST return an HTTP 400 error
to the client.
The issuer creates a BatchTokenResponse structured as follows:

~~~tls

> struct {
>   TokenResponse token_responses<0..2^16-1>;
> } BatchTokenResponse

Reviewer 2: I think you'll need some syntax changes to support this. How about the following?

```suggestion
struct {
  TokenResponse token_response<0..2^16-1>; /* Defined by token_type */ 
} InnerTokenResponse;

struct {
  InnerTokenResponse token_responses<0..2^16-1>;
} BatchTokenResponse
```

You'd also need to say that a length of 0 for `token_response` indicates it's empty (skipped), where a non-empty length means it's not skipped.

Author: i've updated the wording around skip. it was included below, but not clear. I've moved it to the structure definition.

I've used your definition for `InnerTokenRequest`, but without `<0..2^16-1>`, the it is not an array. Maybe we could use opaque again here, but the comment make this clear.

~~~

If the FinalizeBatch function fails, the Client aborts the protocol. Token
verification works exactly as specified in {{RFC9578}}.

# Arbitrary Batched Token Issuance

This section describes an issuance protocol mechanism for issuing multiple tokens in one round trip between Client and Issuer. An arbitrary batched token request can contain token requests for any token type.

## Client-to-Issuer Request {#arbitrary-client-to-issuer-request}

The Client first creates all TokenRequest it wants to batch. To do so, the client follows
protocol describing issuance, such as {{RFC9578, Section 5.1}} or {{RFC9578, Section 6.1}}.

The Client then creates a BatchedTokenRequest structured as follows:

~~~tls
struct {
   uint16_t token_type;
   select (token_type) {
      case (0x0001): /* Type VOPRF(P-384, SHA-384), RFC 9578 */
         uint8_t truncated_token_key_id;
         uint8_t blinded_msg[Ne];
      case (0x0002): /* Type Blind RSA (2048-bit), RFC 9578 */
         uint8_t truncated_token_key_id;
         uint8_t blinded_msg[Nk];
   }

> } TokenRequest;

Reviewer: It would be good to explain how this would work for token types that are not yet defined

struct {
  TokenRequest token_requests<0..2^16-1>;
} BatchTokenRequest
~~~

The structure fields are defined as follows:

- "token_type" is a 2-octet integer. TokenRequest MUST be prefixed with a uint16 "token_type" indicating the token type. The rest of the structure follows based on that type, within the inner opaque token_request attribute. The above definition corresponds to TokenRequest from {{RFC9578}}

- "token_requests" are serialized TokenRequests, in network byte order. The number of token_requests, as a 2-octet integer, is prepended to the serialized TokenRequests. In addition, the 2-octet integer length of each TokenRequest is prepended to the serialized TokenRequests.

The Client then generates an HTTP POST request to send to the Issuer Request
URL, with the BatchTokenRequest as the content. The media type for this request
is "application/private-token-batch-request". An example request for the Issuer
Request URL "https://issuer.example.net/request" is shown below.

~~~
POST /request HTTP/1.1
Host: issuer.example.net

> Accept: application/private-token-batch-response

Reviewer: As discussed offline, we need 4 media types: 

 - private-token-privately-verifiable-batch-{request|response}
 - private-token-arbitrary-batch-{request|response}

   }
} TokenRequest;
~~~

The Client then generates an HTTP POST request to send to the Issuer Request
URL, with the BatchTokenRequest as the content. The media type for this request
is "application/private-token-batch-request". An example request for the Issuer
Request URL "https://issuer.example.net/request" is shown below.

~~~
POST /request HTTP/1.1
Host: issuer.example.net
Accept: application/private-token-batch-response
Content-Type: application/private-token-batch-request
Content-Length: <Length of BatchTokenRequest>

<Bytes containing the BatchTokenRequest>
~~~

## Issuer-to-Client Response {#arbitrary-issuer-to-client-response}

Upon receipt of the request, the Issuer validates the following conditions:

- The Content-Type is application/private-token-batch-request as registered with IANA.

If this condition is not met, the Issuer MUST return an HTTP 400 error
to the client.

The Issuer then tries to deserialize the first 2 bytes of the i-th element
of BatchTokenRequest.token_requests.
If this is not a token type registered with IANA, the Issuer MUST return an HTTP 400 error
to the client.
The issuer creates a BatchTokenResponse structured as follows:

~~~tls
struct {

>   TokenResponse token_response; /* Defined by token_type */

Reviewer 2: This isn't right. We need to have a length prefix to indicate if the field actually exists or not.

Author: updated. the text was correct but convoluted, I've updated it as well

The Client then creates a BatchedTokenRequest structured as follows:

~~~tls
struct {
   uint16_t token_type;
   select (token_type) {
      case (0x0001): /* Type VOPRF(P-384, SHA-384), RFC 9578 */
         uint8_t truncated_token_key_id;
         uint8_t blinded_msg[Ne];
      case (0x0002): /* Type Blind RSA (2048-bit), RFC 9578 */
         uint8_t truncated_token_key_id;
         uint8_t blinded_msg[Nk];
   }
} TokenRequest;

struct {
  TokenRequest token_requests<0..2^16-1>;
} BatchTokenRequest
~~~

The structure fields are defined as follows:

- "token_type" is a 2-octet integer. TokenRequest MUST be prefixed with a uint16 "token_type" indicating the token type. The rest of the structure follows based on that type, within the inner opaque token_request attribute. The above definition corresponds to TokenRequest from {{RFC9578}}

- "token_requests" are serialized TokenRequests, in network byte order. The number of token_requests, as a 2-octet integer, is prepended to the serialized TokenRequests. In addition, the 2-octet integer length of each TokenRequest is prepended to the serialized TokenRequests.

> ~~~tls
> struct {
>    uint16_t token_type;
>    select (token_type) {
>       case (0x0001): /* Type VOPRF(P-384, SHA-384), RFC 9578 */
>          uint8_t truncated_token_key_id;
>          uint8_t blinded_msg[Ne];
>       case (0x0002): /* Type Blind RSA (2048-bit), RFC 9578 */
>          uint8_t truncated_token_key_id;
>          uint8_t blinded_msg[Nk];
>    }
> } TokenRequest;
> ~~~

Reviewer 2: ```suggestion
```

The Client then generates an HTTP POST request to send to the Issuer Request
URL, with the BatchTokenRequest as the content. The media type for this request
is "application/private-token-batch-request". An example request for the Issuer
Request URL "https://issuer.example.net/request" is shown below.

~~~
POST /request HTTP/1.1
Host: issuer.example.net
Accept: application/private-token-batch-response
Content-Type: application/private-token-batch-request
Content-Length: <Length of BatchTokenRequest>

<Bytes containing the BatchTokenRequest>
~~~

## Issuer-to-Client Response {#arbitrary-issuer-to-client-response}

Upon receipt of the request, the Issuer validates the following conditions:

- The Content-Type is application/private-token-batch-request as registered with IANA.

If this condition is not met, the Issuer MUST return an HTTP 400 error
to the client.

The Issuer then tries to deserialize the first 2 bytes of the i-th element
of BatchTokenRequest.token_requests.
If this is not a token type registered with IANA, the Issuer MUST return an HTTP 400 error
to the client.
The issuer creates a BatchTokenResponse structured as follows:

~~~tls
struct {

>   TokenResponse token_response<0..2^16-1>; /* Defined by token_type */
> } InnerTokenResponse;
> 
> struct {
>   InnerTokenResponse token_responses<0..2^16-1>;

Reviewer 2: ```suggestion
  TokenResponse token_response<0..2^16-1>; /* Defined by token_type */
} OptionalTokenResponse;

struct {
  OptionalTokenResponse token_responses<0..2^16-1>;
```

} BatchTokenResponse
~~~

> The structure fields are defined as follows:
> 
> - "token_response" are serialized TokenResponse, in network byte order,
> as defined by their token type. A size 0 indicates that the Issuer
> failed or refused to issue the associated TokenRequest.
> 
> - "token_responses" are serialised InnerTokenResponses. The number of
> token_responses, as a 2-octet integer, is prepended to the serialized
> InnerTokenResponses This matches the number of incoming
> token_requests.

Reviewer 2: ```suggestion
BatchTokenResponse.token_responses is a vector of OptionalTokenResponses, length
prefixed with two bytes. OptionalTokenResponse.token_response is a length-prefix-encoded
TokenResponse, where a length of 0 indicates that the Issuer failed or refused to issue the
associated TokenRequest.
```

The Issuer generates an HTTP response with status code 200 whose content
consists of TokenResponse, with the content type set as
"application/private-token-batch-response".

If the Issuer issues some tokens but not all,
it MUST return an HTTP 206 to the client and continue processing further requests.

~~~
HTTP/1.1 200 OK
Content-Type: application/private-token-batch-response
Content-Length: <Length of BatchTokenResponse>

<Bytes containing the BatchTokenResponse>
~~~

## Finalization {#arbitrary-finalization}

The Client tries to deserialize the i-th element
of BatchTokenResponse.token_responses using the protocol associated to BatchTokenRequest.token_type.

> 

Reviewer 2: ```suggestion
```

If the element has a size of 0, the Client MUST ignore this token, and continue processing the next token.

> 
> If the issuance protocol defines it, the Client finalizes these tokens. If this function fails, the Client aborts the protocol.
> Token verification is unchanged.

Reviewer 2: ```suggestion
The Client finalizes each deserialized TokenResponse using the matching TokenRequest according
to the corresponding finalization procedure defined by the token type.
```

If the FinalizeBatch function fails, the Client aborts the protocol. Token
verification works exactly as specified in {{ISSUANCE}}.

# Arbitrary Batched Verifiable Token

## Client-to-Issuer Request {#arbitrary-client-to-issuer-request}

The Client first creates all TokenRequest it wants to batch. To do so, the client follows
protocol describing issuance, such as {{ISSUANCE, Section 5.1}} or {{ISSUANCE, Section 6.1}}.

The Client then creates a BatchedTokenRequest structured as follows:

~~~tls
struct {

>   uint16_t token_type;

Author: [RFC 9576](https://www.rfc-editor.org/rfc/rfc9576#section-3.5-2) specifies that
> The structure and semantics of the TokenRequest and TokenResponse messages depend on the issuance protocol and token type being used

This means that while all existing TokenRequest start with `token_type`, it's a convention and might break in future Issuance protocols.

Therefore the restriction to one token type per batched request, instead of prepending token_type to all TokenRequest. This also seems like a way to fail fast for an issuer, which would know if they can support the batched requests, knowing both the number of request and their type.

>   uint16_t token_type;

Reviewer 2: That applies to the application/private-token-request media type. But if we use a different media type, as we've done here, that restriction no longer applies.

>   uint16_t token_type;

Author: Would restricting this new issuance protocol to TokenRequest that use `{ uint16_t token_type, ... }` be a possibility? It limits the future TokenRequest, if they want to use Arbitrary Batching.

>   uint16_t token_type;

Author: the proposal has been updated to reflect the comment just above.

  TokenRequest token_requests<0..2^16-1>;
} BatchedTokenRequest
~~~

The structure fields are defined as follows:

- "token_type" is a 2-octet integer, which matches the type of the TokenRequests in the batch.

- "token_requests" are serialized TokenRequests, in network byte order. The number of token_requests, as a 2-octet integer, is prepended to the serialized TokenRequests. In addition, the 2-octet integer length of each TokenRequest is prepended to the serialized TokenRequests.

> The Client MUST ensure all token_requests have the same token_type, which is the same as the BatchedTokenRequest.token_type.

Author: The deserialisation depends on the token type. This requirement is meant to ensure token request are valid.

> The Client MUST ensure all token_requests have the same token_type, which is the same as the BatchedTokenRequest.token_type.

Reviewer 2: I'm suggesting deleting the requirement that each token request in the batch is for the same token type.

> The Client MUST ensure all token_requests have the same token_type, which is the same as the BatchedTokenRequest.token_type.

Author: I'll remove this requirement while adding the ability to fail on specific requests

The Client then generates an HTTP POST request to send to the Issuer Request
URL, with the BatchTokenRequest as the content. The media type for this request
is "application/private-token-batch-request". An example request for the Issuer
Request URL "https://issuer.example.net/request" is shown below.

~~~
POST /request HTTP/1.1
Host: issuer.example.net
Accept: application/private-token-batch-response
Content-Type: application/private-token-batch-request
Content-Length: <Length of BatchTokenRequest>

<Bytes containing the BatchTokenRequest>
~~~

## Issuer-to-Client Response {#arbitrary-issuer-to-client-response}

Upon receipt of the request, the Issuer validates the following conditions:

- The BatchTokenRequest contains a supported token_type registered with IANA.

> - The BatchTokenRequest contains token_requests with the same token_type, matching BatchTokenRequest.token_type.

Author: see comment above. might be revised depending on the outcome.

> - The BatchTokenRequest contains token_requests with the same token_type, matching BatchTokenRequest.token_type.

Author: removed

If any of these conditions is not met, the Issuer MUST return an HTTP 400 error
to the client.
For performance, an Issuer MAY return an HTTP 400 error if all token_requests are

> not targetted at the same cryptographic material.

Author: The rational here is an issuer might allow batched token, but only if it's the same cryptographic key for performance and security reason. The spec highlighting this fact makes it clear that's a possibility.

> not targetted at the same cryptographic material.

Author: removed

The Issuer then tries to deserialize the i-th element
of BatchTokenRequest.token_requests using the protocol associated to BatchTokenRequest.token_type.
If this fails for any of
the BatchTokenRequest.token_requests values, the Issuer MUST return an HTTP 400
error to the client. Otherwise, if the Issuer is willing to produce a token to
the Client, the issuer creates a BatchTokenResponse structured as follows:

~~~tls
struct {

>   TokenResponse token_responses<0..2^16-1>;

Author: The proposal here is to fail.

If we'd like to support multiple requests, some succeeding, some failing, having the client provide an index for each request seems like a good solution.
This can be added to `BlindedElement`

>   TokenResponse token_responses<0..2^16-1>;

Author: I've updated the data structure so that I define a `TokenRequest` with `token_type` and an opaque element

>   TokenResponse token_responses<0..2^16-1>;

Author: for a first pass, this has been updated as follow:

issuer reads token_requests as they arrive. if the i-th fails or is not supports, they return a zero-sized element.

in addition, token_request have token_type as their first element. this is enforced by the issuer.

} BatchedTokenResponse
~~~

The structure fields are defined as follows:

- "token_responses" are serialized TokenResponses, in network byte order. The number of token_responses, as a 2-octet integer, is prepended to the serialized TokenResponses This matches the number of incoming token_requests. In addition, the 2-octet integer length of each TokenResponse is prepended to the serialized TokenResponses.

The Issuer generates an HTTP response with status code 200 whose content
consists of TokenResponse, with the content type set as
"application/private-token-batch-response".

~~~
HTTP/1.1 200 OK
Content-Type: application/private-token-batch-response
Content-Length: <Length of BatchTokenResponse>

<Bytes containing the BatchTokenResponse>
~~~

An Issuer MAY accept streamed request, and MAY provide TokenResponse to the Client as they are ready.

## Finalization {#arbitrary-finalization}

The Client MUST validate that the TokenResponse.token_type matches the

> BatchedTokenRequest.token_type.

Author: removed

The Client then tries to deserialize the i-th element
of BatchTokenResponse.token_responses using the protocol associated to BatchTokenRequest.token_type.
It then finalizes these tokens as specified in their respective protocols.

> If the FinalizeBatch function fails, the Client aborts the protocol. Token

Author: updated the phrasing here.

verification is unchanged.

# Security considerations {#security-considerations}

## Batched Privately Verifiable Tokens

Implementors SHOULD be aware of the security considerations described in {{OPRF,
Section 6.2.3}} and implement mitigation mechanisms. Application can mitigate
this issue by limiting the number of clients and limiting the number of token
requests per client per key.

## Arbitrary Batched Verifiable Tokens

Implementors SHOULD be aware of the inherent linear cost of this token type.

> Applications SHOULD place limits on the number of tokens per request.

Reviewer 2: How do they impose this limit? I think they should have the ability to skip or ignore token requests if they're too many.

Author: I've updated this line to mention Issuer might skip token requests
